File: src/contests/top150/n124_binary_tree_max_path_sum/Solution.java
package contests.top150.n124_binary_tree_max_path_sum;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}


public class Solution {

    private int max = Integer.MIN_VALUE;

    public int maxPathSum(TreeNode root) {
        helper(root);
        return max;
    }

    private int helper(TreeNode node) {
        if (node == null) return 0;

        int left = Math.max(0, helper(node.left));
        int right = Math.max(0, helper(node.right));

        max = Math.max(max, node.val + left + right);
        return Math.max(left, right) + node.val;
    }
}

File: src/contests/top150/n124_binary_tree_max_path_sum/solution.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func maxPathSum(root *TreeNode) int {
	maximum := -1001
	dfs(root, &maximum)
	return maximum
}

func dfs(root *TreeNode, maximum *int) int {
	if root == nil {
		return 0
	}

	left := max(0, dfs(root.Left, maximum))
	right := max(0, dfs(root.Right, maximum))

	*maximum = max(*maximum, root.Val+left+right)
	return max(left, right) + root.Val
}
